syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,987 @@
+
+    @with_session
+    def get_total_size(self, session, user, recalculate=False):
+        """Returns the total size a user's stored data."""
+        userid = user_key(user)
+        size = session.transaction.execute_sql(
+            STORAGE_SIZE,
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        ).one_or_none()
+        return int(size[0] if size else 0)
+
+    @with_session
+    def delete_storage(self, session, user):
+        """Removes all data for the user."""
+        userid = user_key(user)
+
+        # Deleting user_collections will wipe out all BSO's and batches
+        # for the user due to cascading delete
+        session.transaction.execute_update(
+            getq(queries.DELETE_ALL_COLLECTIONS),
+            params={"userid": userid},
+            param_types={"userid": param_types.STRING}
+        )
+
+    #
+    # APIs to operate on an individual collection
+    #
+
+    @with_session
+    def get_collection_timestamp(self, session, user, collection):
+        """Returns the last-modified timestamp of a collection."""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        # The last-modified timestamp may be cached on the session.
+        if session.collection_ts:
+            return session.collection_ts
+
+        # Otherwise we need to look it up in the database.
+        result = session.transaction.execute_sql(
+            getq(queries.COLLECTION_TIMESTAMP),
+            params={"userid": userid,
+                    "collectionid": collectionid},
+            param_types={"userid": param_types.STRING,
+                         "collectionid": param_types.INT64}
+        ).one_or_none()
+        if result is None:
+            raise CollectionNotFoundError
+        return dt2ts(result[0])
+
+    @with_session
+    def get_items(self, session, user, collection, **params):
+        """Returns items from a collection."""
+        return self._find_items(session, user, collection, **params)
+
+    @with_session
+    def get_item_ids(self, session, user, collection, **params):
+        """Returns item ids from a collection."""
+        # Select only the fields we need, including those that might
+        # be used for limit/offset pagination.
+        params["fields"] = ["id", "modified", "sortindex"]
+        res = self._find_items(session, user, collection, **params)
+        res["items"] = [item["id"] for item in res["items"]]
+        return res
+
+    def _find_items(self, session, user, collection, **params):
+        """Find items matching the given search parameters."""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        bind = {"userid": userid,
+                "collectionid": collectionid}
+        bind_types = {"userid": param_types.STRING,
+                      "collectionid": param_types.INT64}
+        ts = get_timestamp()
```

Comment:
If available, should this use a timestamp from the `session` object for consistency?